apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -0,0 +1,702 @@
+        end,
+        Workers
+    ).
+
+%%================================================================================
+%% behavior callbacks
+%%================================================================================
+
+init([CBM, ShardId, Name, Opts]) ->
+    process_flag(trap_exit, true),
+    Pool = emqx_ds_beamformer_sup:pool(ShardId),
+    gproc_pool:add_worker(Pool, Name),
+    gproc_pool:connect_worker(Pool, Name),
+    PendingTab = ets:new(pending_polls, [duplicate_bag, private, {keypos, #poll_req.key}]),
+    WaitingTab = ets:new(pending_polls, [duplicate_bag, private, {keypos, #poll_req.key}]),
+    S = #s{
+        module = CBM,
+        shard = ShardId,
+        metrics_id = shard_metrics_id(ShardId),
+        name = Name,
+        pending_queue = PendingTab,
+        wait_queue = WaitingTab,
+        pending_request_limit = maps:get(pending_request_limit, Opts, 10_000)
+    },
+    self() ! ?housekeeping_loop,
+    {ok, S}.
+
+handle_call(Req = #poll_req{}, _From, S = #s{pending_queue = PendingTab, wait_queue = WaitingTab}) ->
+    %% TODO: it should be possible to drop requests past deadline
+    %% immediately.
+    NQueued = ets:info(PendingTab, size) + ets:info(WaitingTab, size),
+    case NQueued >= S#s.pending_request_limit of
+        true ->
+            Reply = {error, recoverable, too_many_requests},
+            {reply, Reply, S};
+        false ->
+            ets:insert(S#s.pending_queue, Req),
+            {reply, ok, start_fulfill_loop(S)}
+    end;
+handle_call(_Call, _From, S) ->
+    {reply, {error, unknown_call}, S}.
+
+handle_cast(_Cast, S) ->
+    {noreply, S}.
+
+handle_info(#shard_event{updated_streams = Streams}, S) ->
+    logger:debug("Stream event ~p", [Streams]),
+    {noreply, maybe_fulfill_waiting(S, Streams)};
+handle_info(?fulfill_loop, S0) ->
+    S1 = S0#s{is_spinning = false},
+    S = fulfill_pending(S1),
+    {noreply, S};
+handle_info(?housekeeping_loop, S0) ->
+    S = cleanup(S0),
+    erlang:send_after(1000, self(), ?housekeeping_loop),
+    {noreply, S};
+handle_info(_Info, S) ->
+    {noreply, S}.
+
+terminate(_Reason, #s{shard = ShardId, name = Name}) ->
+    Pool = emqx_ds_beamformer_sup:pool(ShardId),
+    gproc_pool:disconnect_worker(Pool, Name),
+    gproc_pool:remove_worker(Pool, Name),
+    ok.
+
+%%================================================================================
+%% Internal exports
+%%================================================================================
+
+-spec start_link(module(), _Shard, integer(), opts()) -> {ok, pid()}.
+start_link(Mod, ShardId, Name, Opts) ->
+    gen_server:start_link(?MODULE, [Mod, ShardId, Name, Opts], []).
+
+%% @doc RPC target: split the beam and dispatch replies to local
+%% consumers.
+-spec do_dispatch(beam()) -> ok.
+do_dispatch(Beam = #beam{}) ->
+    %% TODO: optimize by avoiding the intermediate list. Split may be
+    %% organized in a fold-like fashion.
```

Comment:
Del. No evidence of this being a bottleneck (it would require an extreme level of request coherence, hard to achieve)